### DISCUSSION TITLE
A question about the reordering of the Tibet astrological signs

### DISCUSSION THREAD

**Discussion by cheonhyeongsim** [harfbuzz/harfbuzz#4673] [Q&A] [answered]
[tibet.zip](https://github.com/harfbuzz/harfbuzz/files/15050334/tibet.zip)

In the uploaded zip, there are two font files, perspectively `times.ttf` and `tibet.ttf`. `times.ttf` is a semi-finished product of my non-commercial fontpack, while `tibet.ttf` is a subset of that which contains only Tibetan, Zanabazar, Soyombo and Marchen script.

> Note that, as a semi-finished font modified from Times New Roman, the font name has not been changed, so do not install it without changing its name while testing.

In Harfbuzz, both fonts can render something like `༢༿༾ (U+0F22 U+0F3F U+0F3E)` correctly with the step `end reordering USE` - `U+0F3F` is reordered to the leftmost by Universal Shaping Engine. However, if I add `U+0F19` to the sequence, i.e. `༢༿༙༾ (U+0F22 U+0F3F U+0F19 U+0F3E)`, only `tibet.ttf` could do the reordering by USE, while `times.ttf` does not do the reordering. That seems to be very strange. Such part of the OpenType Layout is the same, and we can see from Crowbar that all the steps during GSUB before the reordering are also the same.

> Note that, after GPOS, you will see the same actual shape even if the reordering did not happen, because I hacked such sequences by GPOS. However, something like `༢༿༙༣༾ (U+0F22 U+0F3F U+0F19 U+0F23 U+0F3E)` will not be hacked in case there would be some over-hacking. So it may be better to use this sequence while testing.

I doubt that if it is because `times.ttf` contains `script DFLT`. But when I add `script DFLT` to `tibet.ttf` (for testing), it still could reorder correctly. So that should not be an issue related to `script DFLT` or `script tibt`.

Additionally, on Windows 8.1, even `༢༿༾ (U+0F22 U+0F3F U+0F3E)` could not be rendered correctly - it is over-reordered, that not only U+0F3F but also U+0F3E are reordered to the leftmost. But on Windows Server 2016, it becomes correct. So I trust that there was an update of the USE. However, on Windows Server 2016, both fonts can reorder `༢༿༙༣༾ (U+0F22 U+0F3F U+0F19 U+0F23 U+0F3E)` correctly (i.e. something like `ʃ` stands on the leftmost).

> i.e. for the sequence `༢༿༙༣༾ (U+0F22 U+0F3F U+0F19 U+0F23 U+0F3E)`:
> DirectWrite + tibet.ttf : Correctly reordered.
> DirectWrite + times.ttf : Correctly reordered.
> Harfbuzz + tibet.ttf : Correctly reordered.
> Harfbuzz + times.ttf : Not reordered.

My question is that, if Harfbuzz passes such sequences to USE (because Tibetan script is listed in the USE Spec), why the behavior is different from DirectWrite?

**Comment by behdad** [comment]
@dscorbett

**Reply by dscorbett** [reply]
HarfBuzz only reorders glyphs for which `_hb_glyph_info_get_lig_comp` returns 0. Right before the reordering for “༢༿༙༣༾”, the buffer is `[uni0F22=0|uni0F19=0|uni0F3F.1=0|uni0F23=3|uni0F3E.1=3]`. In tibet.ttf at that point, all the glyphs have 0 except `uni0F19`, whereas in times.ttf at that point, all the glyphs have 0 except `uni0F3F.1`.

This discrepancy arises in the 'ccmp' lookup that swaps U+0F19 and U+0F3F. Here are the relevant parts of the logs. I modified the logger to print non-0 ligature properties after `##`.
<details>
<summary>tibet.ttf</summary>

```
trace: start lookup 0 feature 'ccmp'	buffer: [uni0F22=0|uni0F3F=0|uni0F19=0|uni0F23=3|uni0F3E=3]
trace: recursing to lookup 1 at 0	buffer: [uni0F22=0|uni0F3F=0|uni0F19=0|uni0F23=3|uni0F3E=3]
trace: ligating glyphs at 0,2	buffer: [uni0F22=0|uni0F3F=0|uni0F19=0|uni0F23=3|uni0F3E=3]
trace: ligated glyph at 0	buffer: [uni0F22=0|uni0F3F=0|uni0F23=3|uni0F3E=3]
trace: recursed to lookup 1	buffer: [uni0F22=0|uni0F3F=0|uni0F23=3|uni0F3E=3]
trace: recursing to lookup 2 at 0	buffer: [uni0F22=0|uni0F3F=0|uni0F23=3|uni0F3E=3]
trace: multiplying glyph at 0	buffer: [uni0F22=0|uni0F3F=0|uni0F23=3|uni0F3E=3]
trace: multiplied glyphs at 0,1	buffer: [uni0F22=0|uni0F19=0##1|uni0F3F=0|uni0F23=3|uni0F3E=3]
trace: recursed to lookup 2	buffer: [uni0F22=0|uni0F19=0##1|uni0F3F=0|uni0F23=3|uni0F3E=3]
trace: end lookup 0 feature 'ccmp'	buffer: [uni0F22=0|uni0F19=0##1|uni0F3F=0|uni0F23=3|uni0F3E=3]
```
</details>
<details>
<summary>times.ttf</summary>

```
trace: start lookup 517 feature 'ccmp'	buffer: [uni0F22=0|uni0F3F=0|uni0F19=0|uni0F23=3|uni0F3E=3]
trace: recursing to lookup 518 at 0	buffer: [uni0F22=0|uni0F3F=0|uni0F19=0|uni0F23=3|uni0F3E=3]
trace: ligating glyphs at 0,2	buffer: [uni0F22=0|uni0F3F=0|uni0F19=0|uni0F23=3|uni0F3E=3]
trace: ligated glyph at 0	buffer: [uni0F22=0|uni0F3F=0##1|uni0F23=3|uni0F3E=3]
trace: recursed to lookup 518	buffer: [uni0F22=0|uni0F3F=0##1|uni0F23=3|uni0F3E=3]
trace: recursing to lookup 519 at 0	buffer: [uni0F22=0|uni0F3F=0##1|uni0F23=3|uni0F3E=3]
trace: multiplying glyph at 0	buffer: [uni0F22=0|uni0F3F=0##1|uni0F23=3|uni0F3E=3]
trace: multiplied glyphs at 0,1	buffer: [uni0F22=0|uni0F19=0|uni0F3F=0##1|uni0F23=3|uni0F3E=3]
trace: recursed to lookup 519	buffer: [uni0F22=0|uni0F19=0|uni0F3F=0##1|uni0F23=3|uni0F3E=3]
trace: end lookup 517 feature 'ccmp'	buffer: [uni0F22=0|uni0F19=0|uni0F3F=0##1|uni0F23=3|uni0F3E=3]
```
</details>

The shaper seems to be ligating and multiplying the same glyphs in the same order, so I don’t know why it assigns different ligature properties.

**Reply by cheonhyeongsim** [reply]
> The shaper seems to be ligating and multiplying the same glyphs in the same order, so I don’t know why it assigns different ligature properties.

Yeah that seems to be the main issue. I wonder if anyone could identify the root cause of the problem from the source code, that will be helpful for us font developers.

**Reply by cheonhyeongsim** [reply]
> This discrepancy arises in the 'ccmp' lookup that swaps U+0F19 and U+0F3F.

The main purpose here is to ensure that U+0F19 comes before U+0F18, or something like `༢༘༙` will not be able to provide the correct glyph. So I first ligate the Tibetan numbers with U+0F19 and then multiply back, to place it straight after the numbers. It is not intentional to swap it with U+0F3F. The code will be more complicated if I replace U+0F19 only when U+0F18 exists before it.

I think that my approach itself does not have any problems, because it is not the root cause of why the shaper assigns different ligature properties.

Btw, may I know whether this difference comes from Harfbuzz or USE?

**Reply by cheonhyeongsim** [reply]
I will try to move the swapping part from `ccmp` to `psts` as a workaround, to see whether the problem disappears or not. Anyway, back to the original topic, if the issue itself comes from Harfbuzz, I hope someone could fix it in the newer versions.

**Reply by behdad** [reply]
The `##1` ligature prop is just a "component" value, saying this is the first or second component of a multiple substitution. The actual ligature-id is in the higher bytes, and that's still 0.  That's my understanding at least.

**Reply by cheonhyeongsim** [reply]
But the main question here is that why it assigns `##1` to the different glyph.

**Reply by cheonhyeongsim** [reply, marked as answer]
> I will try to move the swapping part from ccmp to psts as a workaround, to see whether the problem disappears or not.

Fortunately, the bug disappeared.
